bpo-31861: Add operator.aiter and operator.anext

jab opened the pull request:
...along with passing tests. (I added the tests to `test_asyncgen.py` rather than to `test_operator.py` because they fit much better there.)

/cc @1st1 @njsmith @JelleZijlstra et al.

<!-- issue-number: [bpo-31861](https://www.bugs.python.org/issue31861) -->
https://bugs.python.org/issue31861
<!-- /issue-number -->

jab committed:
bpo-31861: Add operator.aiter and operator.anext

asvetlov commented on Lib/operator.py:
```diff
@@ -408,6 +410,55 @@ def ixor(a, b):
 
+# Asynchronous Iterator Operations ********************************************#
+
+async def aiter(*args):
```
`__aiter__` is a regular function. Why is `aiter` async?

asvetlov replied:
Isn't better to avoid `*args`?
Something like
```
_sentinel = object()
def aiter(async_iterable, sentinel=_sentinel): ...
```

jab replied:
Done in 908227b.

asvetlov commented on Lib/operator.py:
```diff
@@ -408,6 +410,55 @@ def ixor(a, b):
+        yield value
+
+
+async def anext(*args):
```
A question about `*args` again

jab replied:
Done in 908227b.

asvetlov commented on Lib/operator.py:
```diff
@@ -408,6 +410,55 @@ def ixor(a, b):
+        raise TypeError(f'anext expected 1 or 2 arguments, got {lenargs}')
+    ait = args[0]
+    if not isinstance(ait, AsyncIterator):
+        raise TypeError(f'anext expected an AsyncIterable, got {type(ait)}')
```
`an AsyncIterator` here

jab replied:
Thanks for catching, done in 908227b.

jab committed:
address comments from first review

jab commented on Lib/test/test_asyncgen.py:
```diff
@@ -386,6 +387,80 @@ def tearDown(self):
+    def test_async_gen_operator_aiter_class(self):
+        loop = self.loop
+        class Gen:
+            async def __aiter__(self):
```
I based this off the `class Gen` defined in the existing `test_async_gen_asyncio_03()` test. But based on your previous comment, should these both use `def` rather than `async def`?

1st1 replied:
1. `__aiter__` must be **a synchronous function** (`def`) and must return *an asynchronous iterator*.

2. Whan `__aiter__` is declared as an asynchronous generator, it is still **a synchronous function**, that just returns an instance of *an asynchronous generator* when you call it (so (1) still stands).

jab replied:
Thanks for explaining, but I'm still confused. After reading (1) it sounded like `__aiter__` should be rewritten, but after reading (2) it sounded like the code should stay written as-is (even though it seems more confusing this way).

1st1 replied:
It's OK for it to stay written as is.

> After reading (1) it sounded like __aiter__ should be rewritten, but after reading (2) it sounded like the code should stay written as-is (even though it seems more confusing this way).

(2) is an edge case of (1); the case of `__iter__` returning an iterator and `__iter__` being a generator function is absolutely the same.

1st1 reviewed (changes requested):
Please don't merge this yet. I'm not convinced that aiter and anext shouldn't be builtins.

nanjekyejoannah reviewed (commented):
I think I agree, this is good feedback for @ikrivosheev but anyone can pick up these suggestions in a different PR unless @ikrivosheev commits again.

sweeneyde commented on Lib/operator.py:
```diff
@@ -408,6 +410,56 @@ def ixor(a, b):
+            raise TypeError(f'aiter expected an AsyncIterable, got {type(obj)}')
+        if isinstance(obj, AsyncIterator):
+            return obj
+        return (i async for i in obj)
```
```suggestion
        ait = obj.__aiter__()
        if not hasattr(ait, '__anext__'):
            raise TypeError(f"aiter() returned non-AsyncIterator of type {type(ait)!r}")
        return ait
```
I think it should be this way for a few reasons:

1. It eliminates the two separate non-raising cases and return statements/types
2. It eliminates the overhead of the extra async generator expression and is more easily implemented in C.
4. For any AsyncIterator `a`, we "should" have `a.__aiter__() is a`, so the behavior shouldn't change much.
5. For the cases where that isn't true, in analogy with the builtin `iter()`, one might expect `aiter(x)` to always call `x.__aiter__()`:

```python
from collections.abc import Iterator, AsyncIterator

class A:
    def __next__(self):
        return 1
    def __iter__(self):
        return iter([])

# "a" is already an iterator, but a.__iter__ is called anyway.
a = A()
assert isinstance(a, Iterator)
assert iter(a) is not a
assert type(iter(a)) is not type(a)
assert type(iter(a)) is type(iter([]))

############################

async def another_aiter():
    yield 1

class B:
    def __anext__(self):
        return 1
    def __aiter__(self):
        return another_aiter()

b = B()
assert isinstance(b, AsyncIterator)
# These three assertions fail, but pass with the change.
assert aiter(b) is not b
assert type(aiter(b)) is not type(b)
assert type(aiter(b)) is type(aiter(another_aiter()))
```

See the implementation of `iter()` [here](https://github.com/python/cpython/blob/master/Objects/abstract.c#L2616), which uses `PyObject_GetIter()` from [here](https://github.com/python/cpython/blob/master/Python/bltinmodule.c#L1511).
I wonder if a `PyObject_GetAsyncIter` needs to be added to the C-API so that `aiter()` and [the GET_AITER opcode](https://github.com/python/cpython/blob/master/Python/ceval.c#L2196) can use the same code.

JelleZijlstra replied:
Shouldn't we call `type(obj).__aiter__(obj)` (and the analogue for `__anext__`)? That's how dunder methods are generally called by the interpreter.

jab replied:
Thanks for reviewing! Updated this patch based on the good suggestions.

(Note, I used `isinstance(ait, AsyncIterator)` rather than the `hasattr(ait, '__anext__')` that was suggested, figuring that's the preferred spelling. Also, after rebasing on latest master, I had to move the `from collections.abc ...` imports from module scope to function scope as a workaround to avoid a circular dependency between `operator` and `collections`, which wasn't an issue when I first submitted this PR. Sounds like this PR will now serve as a reference implementation for a reimplementation in C (submitted in a separate PR) anyway, where I guess such a workaround won't be necessary?)

jab committed:
Address new review comments + rebase

...on top of latest master.

Also drop now-removed `loop` kwarg from asyncio.sleep call.
Ref: https://bugs.python.org/issue42392

jab committed:
improve tests
